### PR TITLE
Featur: HITS

### PR DESCRIPTION
**Pull request by jingyimei**
JIRA:MADLIB-1124

Introducees a new module that computes the HITS scores of all nodes in a
directed graph.
Implements the HITS algorithm with normalization
(https://en.wikipedia.org/wiki/HITS_algorithm)

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/196/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/198/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/201/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/205/

File: src/ports/postgres/modules/graph/hits.py_in [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+    Returns:
+        String. Help/usage information
+    """
+    if message is not None and \
+            message.lower() in ("usage", "help", "?"):
+        help_string = "Get from method below"
+        help_string = get_graph_usage(schema_madlib, 'HITS',
+                                      """out_table     TEXT, -- Name of the output table for HITS
+    max_iter      INTEGER, -- Maximum iteration number (DEFAULT = 100)
+    threshold     DOUBLE PRECISION, -- Stopping criteria (DEFAULT = 1/(N*1000),
```
Comment (njayaram2) [inline src/ports/postgres/modules/graph/hits.py_in]: Indent an extra tab (4 spaces) to the right, for both `max_iter` and `threshold`.

File: src/ports/postgres/modules/graph/hits.py_in [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+        ##########################################################################
+        # HITS scores for node 'A' at any given iteration 'i' is calculated as following:
+        # authority_i(A) = hub_i(B) + hub_i(C) + ...
+        # After calculating authority score for all nodes, a normalization will 
+        # be done for authority_i(A), authority_i(B), authority_i(C), ... 
+        # Then
+        # hub_i(A) = authority_i(B) + authority_i(C) + ... 
+        # After calculating hub score for all nodes, a normalization will 
+        # be done for hub_i(A), hub_i(B), hub_i(C), ...
+        # B, C are nodes that have edges that point to node A
```
Comment (njayaram2) [inline src/ports/postgres/modules/graph/hits.py_in]: Can you please double check the algorithm? It looks like the order in which you perform updates and normalize must be different. I checked the source code in Wikipedia and it is similar to what you have done, but the original paper seems to be different. Look at page 9 in https://www.cs.cornell.edu/home/kleinber/auth.pdf.
```
auth_{i} <- USE hub_{i-1} and auth_{i-1}
hub_{i} <- USE hub_{i-1} and auth_{i}
auth_{i} <- normalize(auth_{i})
hub_{i} <- normalize(hub_{i})
```

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/220/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/221/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/225/

**Comment by njayaram2** [comment]
Jenkins ok to test

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/226/

File: src/ports/postgres/modules/graph/hits.py_in [hunk 3]
```diff
@@ -0,0 +1,417 @@
...
+                CREATE TABLE {summary_table} (
+                    __iterations__ INTEGER
+                )
+            """.format(**locals()))
+
+        unconverged_authority_num = 0
+        unconverged_hub_num = 0
+        iteration_num = 0
+        authority_norm = 0
+        hub_norm = 0
```
Comment (njayaram2) [inline src/ports/postgres/modules/graph/hits.py_in]: `hub_norm` and `authority_norm` can be initialized to `1`.

File: src/ports/postgres/modules/graph/hits.py_in [hunk 4]
```diff
@@ -0,0 +1,417 @@
...
+
+            # normalize hub value with L2 distance
+            hub_norm = math.sqrt(plpy.execute("""
+                    SELECT SUM({tmp3}.hub_square) AS sum_hub_square
+                    FROM
+                    (SELECT POWER(hub, 2) AS hub_square
+                    FROM {message}) AS {tmp3}
+                """.format(**locals()))[0]["sum_hub_square"])
+
+            if hub_norm != 0:
```
Comment (njayaram2) [inline src/ports/postgres/modules/graph/hits.py_in]: `hub_norm == 0` and `authority_norm == 0` are error cases. You could probably check for `if hub_norm == 0 or authority_norm == 0:` and exit the program with an error message on True.

File: src/ports/postgres/modules/graph/hits.sql_in [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+  0 | 0.0865332738776 |  0.375721659592658
+  1 | 0.1838832069899 |  0.533118571043637
+  2 |  0.432666369388 |   0.65497424442504
+  3 | 0.7030828502555 | 0.0406185577938009
+  4 | 0.1838832069899 | 0.0406185577938009
+  5 | 0.3028664585716 |  0.182783510072104
+  6 | 0.3893997324492 |  0.330025782074632
+(7 rows)
+</pre>
+<pre class="result">
```
Comment (njayaram2) [inline src/ports/postgres/modules/graph/hits.sql_in]: Must add the `SELECT * FROM hits_out_summary` query.

File: src/ports/postgres/modules/graph/hits.sql_in [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+  0 | 0.00732986181947514 |   0.351368354456342
+  1 |   0.167097943665847 |   0.530913722392824
+  2 |   0.416198716437073 |   0.668522269025145
+  3 |   0.717639283762987 | 0.00343216951849906
+  4 |   0.167097943665847 | 0.00343216951849906
+  5 |   0.311633656418623 |    0.18657058991966
+  6 |   0.398446707343031 |   0.336030846920549
+(7 rows)
+</pre>
+<pre class="result">
```
Comment (njayaram2) [inline src/ports/postgres/modules/graph/hits.sql_in]: Must add the `SELECT * FROM hits_out_summary` query.

File: src/ports/postgres/modules/graph/hits.sql_in [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+<pre class="syntax">
+DROP TABLE IF EXISTS hits_out, hits_out_summary;
+SELECT madlib.hits(
+             'vertex',             -- Vertex table
+             'id',                 -- Vertex id column
+             'edge',               -- Edge table
+             'src=src, dest=dest', -- Comma delimited string of edge arguments
+             'hits_out',           -- Output table of HITS
+             NULL,
+             0.01);                   -- Max iteration
```
Comment (njayaram2) [inline src/ports/postgres/modules/graph/hits.sql_in]: `Max iteration`, or `threshold`?
Please have the indentation aligned in the SQL query comments too whenever possible.

File: src/ports/postgres/modules/graph/hits.py_in (hunk 4 above)
Comment (jingyimei) [reply src/ports/postgres/modules/graph/hits.py_in]: Yes, this makes more sense.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/229/

**Comment by jingyimei** [comment]
Jenkins ok to test

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/230/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/231/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/234/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/236/

**Comment by jingyimei** [comment]
Jenkins ok to test

**Comment by jingyimei** [comment]
Jenkins ok to test

**Comment by jingyimei** [comment]
Jenkins ok to test

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/237/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/238/

File: src/ports/postgres/modules/graph/hits.sql_in [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+    single model is generated for all data.
+    @note Expressions are not currently supported for 'grouping_cols'.
+    The grouping support will be added later.</dd>
+
+</dl>
+
+@anchor notes
+@par Notes
+
+1. The HITS algorithm is based on Kleinburg's paper [1].
```
Comment (njayaram2) [inline src/ports/postgres/modules/graph/hits.sql_in]: `Kleinburg` -> `Kleinberg`

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/242/

**Pull request closed without merging**